# PathogenSurveillance Workshop

This session is part of [**Nanopore Sequencing and Whole Genome Assembly and Annotation with nf-core/pathogensurveillance**](https://events.rdmobile.com/Sessions/Details/2316143)

---

## Table of Contents

1. [Resources](#resources)
2. [Prerequisites](#prerequisites)
3. [Setting up Gitpod](#setting-up-gitpod)
X. [Session Leader(s) Bios](#session-leaders-bios)
---

## Resources

- [PathogenSurveillance Repository](https://github.com/grunwaldlab/pathogensurveillance)
- [PSMiner Repository](https://github.com/grunwaldlab/psminer)

---

## Prerequisites

- **Some familiarity with the Linux command-line interface**  
  If not, look here for resources:
  - [Command Line and Filesystem](https://open.oregonstate.education/computationalbiology/chapter/the-command-line-and-filesystem/)
  - [PCfB Appendices](content/PCfB_Appendices.pdf)

- **GitHub account**  
  If you don't have one, please create an account at [GitHub](https://github.com):
  - [Creating an Account on GitHub](https://docs.github.com/en/get-started/start-your-journey/creating-an-account-on-github)

- **Understanding of genomic terms**:
  - Reads
  - Genome
  - Populations
  - Contigs
  - Annotations
  - Phylogenies
  - Variants/SNPs
  - Homology

- **Understanding of the need for bioinformatics workflows**

- **Understanding the general role and input/output files**

---

## Setting up Gitpod

### What is Gitpod?

We are using *Gitpod* to provide hands-on experience with installing and using the pathogen surveillance pipeline we developed. [**Gitpod**](https://gitpod.io/) allows you to launch and enter Virtual Machines from your browser and gives you 50 free hours per calendar month, enough for this session.

### How do I sign up for Gitpod?

You MUST do the following steps before the workshop starts! It will take a few minutes to set up a **GitHub** account and a **Gitpod** account if you don't already have them:

### 1. **Step 00**  
   Click on [this link](https://gitpod.io/new#https://github.com/grunwaldlab/aps_workshop) in any desktop web browser (**Note: Safari sometimes has problems on a Mac, so use Chrome or Mozilla instead)**  
   ![Step 00](content/Untitled%206.png)

### 2. **Step 01**  
   It will ask you to sign up with a GitHub account. If you don't have one, please create an account at [GitHub](https://github.com).  
   ![Step 01](content/Untitled%207.png)

### 3. **Step 02**  
   Once you are signed in to GitHub, click on the Gitpod link above again. It will try to launch our Gitpod workspace using your GitHub login. You will need to give Gitpod permission to access (only) the email address on your GitHub account. Follow the Gitpod prompts to ensure you are a real human and won't misuse their resources. **Note: if it asks you to connect your LinkedIn, you can skip that - you will still get 50 hours :-)**

### 4. **Step 03**  
   Gitpod will ask you to select which code editor and which machine size to start. The defaults are fine. So just click **Continue**  
   ![Step 03](content/Untitled%208.png)

### 5. **Step 04**  
   Once the Gitpod workspace starts (typically in less than a minute), you will see the following in your desktop web browser: a file browser on your left, a text editor top right, and a Linux terminal bottom right.  
   ![Step 04](content/Untitled.jpeg)

> *If you are doing this before the workshop, please go to [gitpod.io/workspaces](https://gitpod.io/workspaces) to delete any running workspaces (in green) so that you don't use up your 50 hours per month* 🙂

---

# Exploring PathogenSurveillance Outputs

This document will guide you through the key output directories generated by the `PathogenSurveillance` pipeline. You will learn where to find the outputs and understand their significance. This section is designed to be viewed in VSCode Gitpod, with the file explorer on the left, this document previewed on the top panel, and the terminal on the bottom panel for an efficient and interactive learning experience.


## `sendsketch`

The `sendsketch` process in the PathogenSurveillance pipeline utilizes the BBMap tool to generate sketch data for the input sequences. This step is crucial for identifying and comparing genomic sequences against a reference database.

### Key output directory: `output/bbmap_sendsketch/`

This directory contains symbolic links to the actual output files generated by the `sendsketch` process. The outputs are stored in the working directory under a unique identifier.

```sh
# Let's list the output for sendsketch:
ll output/bbmap_sendsketch/

# Take a look at output for the sample we included here:
head output/bbmap_sendsketch/*.txt
```

<details>
<summary> 💡 Learn more here </summary>

  ### Inputs
  
  - **Input Sequences**: The process takes in sequencing data files, such as FASTQ files.
  - **Reference Database**: The sequences are compared against a reference database (e.g., RefSeq).
  
  ### Outputs
  
  - **Sketch Files**: These are the primary outputs containing information about the query sequences and their comparison to the reference database.
  
  ### Understanding the Outputs
  
  The main output file `22_323.txt` contains detailed information about the query sequence and its comparison to the reference database. Here are some key fields:
  
  - **Query**: Information about the query sequence, including its ID, run ID, sample ID, and other metadata.
  - **DB**: The reference database used for comparison (e.g., RefSeq).
  - **SketchLen**: The length of the sketch.
  - **Seqs**: Number of sequences in the sketch.
  - **Bases**: Total bases in the sketch.
  - **Quality**: Quality score of the sketch.
  - **Depth**: Depth of the sketch.
  - **Taxonomy**: Taxonomic classification of the sequences.
  
  ### Example Output Snippet
  
  ```
  Query: a4d92ba2-9c26-4548-927d-f10ab976c5fb runid=e8007fcc5055076f3e399f09a077cc1f37b2b8ff sampleid=run4 read=2240 ch=338 start_time=2022-08-04T04:43:00Z model_version_id=2021-11-17_dna_r10.4_minion_promethion_1024_67af0493 barcode=barcode07       DB: RefSeq      SketchLen: 35910   Seqs: 7396      Bases: 88561804 gSize: 5424369  GC: 0.640       Quality: 0.7925 AvgCount: 13.456        Depth: 13.456   File: xan_22-323_nanopore.fastq.gz
  WKID    KID     ANI     SSU     SSULen  Complt  Contam  Contam2 uContam Score   E-Val   Depth   Depth2  Volume  RefHits Matches Unique  Unique2 Unique3 noHit   Length  TaxID   ImgID           gBases  gKmers  gSize   gSeqs   GC      rDiv    qDiv    rSize   qSize   cHits      taxName file    seqName taxonomy
  35.22%  22.41%  96.29%  91.52%  1551    100.00% 20.06%  0.62%   4.03%   6793    0.00e+00        14.05   13.78   113.1   34.20   8047    542     6660    13153   9654    35910   863365  -1      5052399 5014802 4995825 1       0.638   22845   35910   22845   35910   7205       Xanthomonas hortorum pv. carotae str. M081      .       tid|863365|NZ_CM002307.1 Xanthomonas hortorum pv. carotae str. M081 chromosome, whole genome shotgun sequence   sk:Bacteria;p:Proteobacteria;c:Gammaproteobacteria;o:Xanthomonadales;f:Xanthomonadaceae;g:Xanthomonas;s:Xanthomonas hortorum;Xanthomonas hortorum pv. carotae
  ```
 
  ### Column Details
  
  - **WKID**: Weighted K-mer ID, representing the percentage of shared k-mers between the query and reference.
  - **KID**: K-mer ID, showing the percentage of identical k-mers.
  - **ANI**: Average Nucleotide Identity, indicating the similarity between the query and reference sequences.
  - **SSU**: Small Subunit ribosomal RNA percentage identity.
  - **SSULen**: Length of the SSU alignment.
  - **Complt**: Completeness of the match.
  - **Contam**: Contamination level in the match.
  - **uContam**: Unique contamination level.
  - **Score**: Score of the match.
  - **E-Val**: E-value of the match.
  - **Depth**: Depth of coverage for the match.
  - **Volume**: Volume of the match.
  - **RefHits**: Number of hits in the reference database.
  - **Matches**: Number of matches in the query sequence.
  - **Unique**: Unique k-mers in the query sequence.
  - **noHit**: Number of k-mers in the query with no hits in the reference.
  - **Length**: Length of the alignment.
  - **TaxID**: Taxonomic ID of the reference.
  - **ImgID**: IMG ID of the reference.
  - **gBases**: Total bases in the genome.
  - **gKmers**: Total k-mers in the genome.
  - **gSize**: Size of the genome.
  - **gSeqs**: Number of sequences in the genome.
  - **GC**: GC content.
  - **rDiv**: Relative diversity.
  - **qDiv**: Query diversity.
  - **rSize**: Size of the reference.
  - **qSize**: Size of the query.
  - **cHits**: Cumulative hits in the query.

  > ## 🧐 What does all this mean? and Why?   
  > 
  >  ### Sketches
  >  
  >  Sketching is a technique used in bioinformatics to create a compact representation of a sequence. This method reduces the memory and computational requirements for sequence comparison. Sketches are generated using a subset of k-mers (short subsequences of length k) from the original sequence.
  >  
  >  ### K-mers
  >  
  >  - **Definition**: K-mers are contiguous sequences of k nucleotides. For example, in the sequence `AGCT`, the 2-mers (k=2) are `AG`, `GC`, and `CT`.
  >  - **Usage**: K-mers are used to break down large sequences into smaller, manageable pieces. They are essential for tasks such as sequence alignment, error correction, and genome assembly.
  >  - **Sketch generation**: A sketch is generated by selecting a representative subset of k-mers from the sequence. This subset captures the essential characteristics of the sequence, allowing for efficient and accurate comparisons.
  >  
  >  ### Importance in Pathogen Surveillance
  >  
  >  - **Efficient Comparison**: Sketches enable the rapid comparison of genomic sequences against large reference databases, facilitating the identification of pathogens.
  >  - **Data Reduction**: By reducing the size of the data to be compared, sketches make it feasible to analyze large datasets within a reasonable time frame and with limited computational resources.

</details>

---

## `flye`

The `flye` process in the PathogenSurveillance pipeline assembles long-read sequencing data to generate high-quality genome assemblies. This step is crucial for obtaining contiguous sequences that can be used for downstream analysis.

### Key output directory: `output/flye_nanopore/`

This directory contains symbolic links to the actual output files generated by the `flye` process. The outputs are stored in the working directory under a unique identifier.

```sh
# Let's list the output for flye:
ll output/flye_nanopore/

# Take a look at the assembly FASTA file:
less -S output/flye_nanopore/22_323.assembly.fasta.gz | head -n 5

# View the Flye log file:
tail -n 10 output/flye_nanopore/22_323.flye.log | head -n 8
```

<details>
<summary> 💡 Learn more here </summary>

  ### Inputs
  
  - **Input Sequences**: The process takes in long-read sequencing data files.
  - **Mode**: Specifies the type of input data (e.g., `--nano-raw` for raw nanopore reads).
  
  ### Outputs
  
  - **Assembly FASTA File**: Contains the assembled genome sequences.
  - **Graph Files (GFA and GV)**: Represent the assembly graph in different formats.
  - **Assembly Info File**: Provides information about the assembly, such as the total length, number of fragments, and coverage.
  - **Log File**: Contains logs generated during the assembly process.
  - **Parameters File**: Stores the parameters used for the assembly process.
  
  ### Understanding the Outputs
  
  The main output file `22_323.assembly.fasta.gz` contains the assembled genome sequences. The `22_323.assembly_info.txt` file provides detailed statistics about the assembly.
  
  ### Example Output Snippet from `assembly_info.txt`
  
  ```
  [2024-07-24 20:16:31] root: INFO: Assembly statistics:

          Total length:   5281721
          Fragments:      1
          Fragments N50:  5281721
          Largest frg:    5281721
          Scaffolds:      0
          Mean coverage:  34
  ```

  ### Column Details
  
  - **Total length**: Total length of the assembled genome.
  - **Fragments**: Number of fragments in the assembly.
  - **Fragments N50**: N50 value for the fragments, indicating that 50% of the assembly is contained in fragments of this length or longer.
  - **Largest frg**: Length of the largest fragment in the assembly.
  - **Scaffolds**: Number of scaffolds in the assembly.
  - **Mean coverage**: Average coverage of the assembly.

  > ## 🧐 What does all this mean? and Why?   
  > 
  >  ### Genome Assembly
  >  
  >  Genome assembly is the process of reconstructing the original genome from sequencing reads. This is achieved by overlapping reads to form longer contiguous sequences called contigs.
  >  
  >  ### Flye
  >  
  >  - **Purpose**: Flye is a de novo assembler for long reads produced by technologies such as Nanopore and PacBio. It is designed to assemble genomes with high continuity and accuracy.
  >  - **Usage**: Flye can handle various types of long-read data, producing high-quality assemblies suitable for downstream analysis.
  >  - **Output**: The resulting assembly includes contigs, assembly graphs, and detailed assembly statistics, which are crucial for understanding the quality and characteristics of the assembly.
  >  
  >  ### Importance in Pathogen Surveillance
  >  
  >  - **High-Quality Assemblies**: Obtaining high-quality genome assemblies is essential for accurate pathogen identification and characterization.
  >  - **Downstream Analysis**: The assembled genomes can be used for various downstream analyses, such as variant calling, gene annotation, and comparative genomics.

</details>

---

## `quast`

The `quast` process in the PathogenSurveillance pipeline evaluates genome assemblies by providing various quality metrics. This step is essential for assessing the completeness and accuracy of the assembled genomes.

### Key output directory: `output/quast/`

This directory contains symbolic links to the actual output files generated by the `quast` process. The outputs are stored in the working directory under a unique identifier.

```sh
# Let's list the output for quast:
ll output/quast/

# Take a look at the detailed output directory for the sample:
ll output/quast/22_323

# View the QUAST report:
less -S output/quast/22_323/report.tsv | head -n 23
```

<details>
<summary> 💡 Learn more here </summary>

  ### Inputs
  
  - **Consensus Sequences**: The assembled sequences to be evaluated.
  - **Reference Genome (optional)**: The reference genome for comparison.
  - **Annotation GFF (optional)**: The annotation file for the reference genome.
  
  ### Outputs
  
  - **Report Files**: Contain various metrics and statistics about the genome assembly.
  - **HTML Reports**: Interactive reports for visualizing the assembly quality.
  - **PDF Reports**: Printable versions of the assembly quality reports.
  - **Log Files**: Contains logs generated during the QUAST evaluation.
  
  ### Understanding the Outputs
  
  The main output files include `report.tsv`, `report.txt`, and `report.pdf`, which provide detailed statistics about the genome assembly. The HTML reports (`icarus.html` and `report.html`) offer interactive visualizations.

  ### Example Output Snippet from `report.tsv`
  
  ```
  Assembly        22_323.assembly
  # contigs (>= 0 bp)     1
  # contigs (>= 1000 bp)  1
  # contigs (>= 5000 bp)  1
  # contigs (>= 10000 bp) 1
  # contigs (>= 25000 bp) 1
  # contigs (>= 50000 bp) 1
  Total length (>= 0 bp)  5281721
  Total length (>= 1000 bp)       5281721
  Total length (>= 5000 bp)       5281721
  Total length (>= 10000 bp)      5281721
  Total length (>= 25000 bp)      5281721
  Total length (>= 50000 bp)      5281721
  # contigs       1
  Largest contig  5281721
  Total length    5281721
  GC (%)  64.04
  N50     5281721
  N90     5281721
  auN     5281721.0
  L50     1
  L90     1
  # N's per 100 kbp       0.00
  ```

  ### Column Details
  
  - **# contigs**: Number of contigs in the assembly.
  - **Total length**: Total length of all contigs.
  - **Largest contig**: Length of the largest contig.
  - **GC (%)**: GC content percentage.
  - **N50**: N50 value of the assembly, indicating the length of the shortest contig at 50% of the total assembly length.
  - **N90**: N90 value of the assembly, indicating the length of the shortest contig at 90% of the total assembly length.
  - **auN**: The auN statistic provides a summary of the assembly's length distribution.
  - **L50**: Number of contigs whose length sum makes up 50% of the total assembly length.
  - **L90**: Number of contigs whose length sum makes up 90% of the total assembly length.
  - **# N's per 100 kbp**: Number of ambiguous bases (Ns) per 100,000 base pairs.

  > ## 🧐 What does all this mean? and Why?   
  > 
  >  ### Genome Assembly Evaluation
  >  
  >  Evaluating the quality of a genome assembly is crucial for understanding its completeness and accuracy. QUAST provides a comprehensive set of metrics to assess these aspects.
  >  
  >  ### QUAST
  >  
  >  - **Purpose**: QUAST (Quality Assessment Tool for Genome Assemblies) evaluates genome assemblies by comparing them to a reference genome and computing various metrics.
  >  - **Usage**: QUAST can handle multiple types of input data, including assembled sequences, reference genomes, and annotation files.
  >  - **Output**: The resulting reports provide detailed statistics and visualizations, allowing researchers to assess the quality of their assemblies effectively.
  >  
  >  ### Importance in Pathogen Surveillance
  >  
  >  - **Quality Assurance**: Ensuring high-quality genome assemblies is essential for accurate pathogen identification and characterization.
  >  - **Data Interpretation**: The detailed metrics and visualizations help researchers interpret the quality and characteristics of their assemblies, facilitating better decision-making in downstream analyses.

</details>

---

## `bakta`

The `bakta` process in the PathogenSurveillance pipeline annotates genomic sequences, providing detailed information about coding sequences (CDSs), RNA genes, and other genomic features. This step is crucial for understanding the functional elements within the genome.

### Key output directory: `output/bakta_bakta/`

This directory contains symbolic links to the actual output files generated by the `bakta` process. The outputs are stored in the working directory under a unique identifier.

```sh
# Let's list the output for bakta:
ll output/bakta_bakta/

# Take a look at the annotation summary:
head -n 28 output/bakta_bakta/22_323.txt
```

<details>
<summary> 💡 Learn more here </summary>

  ### Inputs
  
  - **FASTA File**: The genomic sequences to be annotated.
  - **Database**: The annotation database used by Bakta.
  - **Proteins (optional)**: Additional protein sequences for annotation.
  - **Prodigal Translation Table (optional)**: Translation table for Prodigal gene prediction.
  
  ### Outputs
  
  - **EMBL File**: Annotated sequence in EMBL format.
  - **FASTA Files (faa, ffn, fna)**: Amino acid, nucleotide, and genomic sequences.
  - **GenBank File (gbff)**: Annotated sequence in GenBank format.
  - **GFF3 File**: General Feature Format for annotations.
  - **TSV Files**: Tab-separated values with detailed annotation information.
  - **TXT File**: Summary of the annotation.
  - **Hypotheticals Files**: Detailed information about hypothetical proteins.
  
  ### Understanding the Outputs
  
  The main output file `22_323.txt` provides a summary of the annotation, including counts of various genomic features such as CDSs, RNAs, and hypothetical proteins.
  
  ### Example Output Snippet from `22_323.txt`
  
  ```
  Sequence(s):
  Length: 5281721
  Count: 1
  GC: 64.0
  N50: 5281721
  N ratio: 0.0
  coding density: 86.2

  Annotation:
  tRNAs: 54
  tmRNAs: 1
  rRNAs: 6
  ncRNAs: 69
  ncRNA regions: 7
  CRISPR arrays: 0
  CDSs: 4359
  pseudogenes: 0
  hypotheticals: 577
  signal peptides: 0
  sORFs: 0
  gaps: 0
  oriCs: 3
  oriVs: 0
  oriTs: 0

  Bakta:
  Software: v1.9.4
  Database: v5.1, light
  ```

  ### Column Details
  
  - **Length**: Length of the sequence.
  - **Count**: Number of sequences.
  - **GC**: GC content percentage.
  - **N50**: N50 value of the sequence.
  - **N ratio**: Ratio of ambiguous bases (Ns).
  - **Coding density**: Percentage of the sequence that is coding.
  - **tRNAs**: Number of transfer RNA genes.
  - **tmRNAs**: Number of transfer-messenger RNA genes.
  - **rRNAs**: Number of ribosomal RNA genes.
  - **ncRNAs**: Number of non-coding RNA genes.
  - **ncRNA regions**: Number of non-coding RNA regions.
  - **CRISPR arrays**: Number of CRISPR arrays.
  - **CDSs**: Number of coding sequences.
  - **Pseudogenes**: Number of pseudogenes.
  - **Hypotheticals**: Number of hypothetical proteins.
  - **Signal peptides**: Number of signal peptides.
  - **sORFs**: Number of small open reading frames.
  - **Gaps**: Number of gaps.
  - **oriCs**: Number of chromosomal origin of replication sites.
  - **oriVs**: Number of viral origin of replication sites.
  - **oriTs**: Number of transfer origin of replication sites.

  > ## 🧐 What does all this mean? and Why?   
  > 
  >  ### Genome Annotation
  >  
  >  Genome annotation is the process of identifying functional elements within a genome, such as genes, RNA sequences, and regulatory regions. This information is critical for understanding the biology of the organism.
  >  
  >  ### Bakta
  >  
  >  - **Purpose**: Bakta is an annotation tool that provides detailed genomic annotations using a comprehensive database.
  >  - **Usage**: Bakta can annotate various types of genomic sequences, producing multiple output formats for different applications.
  >  - **Output**: The resulting annotations include coding sequences, RNA genes, hypothetical proteins, and other genomic features, providing a comprehensive view of the genome's functional elements.
  >  
  >  ### Importance in Pathogen Surveillance
  >  
  >  - **Functional Insights**: Understanding the functional elements within a pathogen's genome is essential for identifying virulence factors, antibiotic resistance genes, and other important traits.
  >  - **Data Utilization**: The annotated genome can be used for various downstream analyses, including comparative genomics, evolutionary studies, and functional characterization.

</details>

---

## `sourmash compare`

The `sourmash compare` process in the PathogenSurveillance pipeline computes the similarity matrix for a set of genomic signatures using Average Nucleotide Identity (ANI) and a k-mer length of 31. This step is crucial for comparing genomic sequences and identifying similarities and differences between them.

### Key output directory: `output/sourmash_compare/`

This directory contains symbolic links to the actual output files generated by the `sourmash compare` process. The outputs are stored in the working directory under a unique identifier.

```sh
# Let's list the output for sourmash compare:
ll output/sourmash_compare/

# Take a look at the similarity matrix in CSV format:
less -S output/sourmash_compare/aps_comp.csv | head -n 10 | cut -f 1-10,61 -d ","
```

<details>
<summary> 💡 Learn more here </summary>

  ### Inputs
  
  - **Genomic Signatures**: The process takes in genomic signature files.
  - **File List (optional)**: A list of signature files to be compared.
  - **Save Options**: Options to save the output matrix in numpy and/or CSV format.
  
  ### Outputs
  
  - **Numpy Matrix**: Contains the computed similarity matrix in numpy format (`*.npy` and `*.npy.labels.txt`).
  - **CSV File**: Contains the computed similarity matrix in CSV format.
  
  ### Understanding the Outputs
  
  The main output file `aps_comp.csv` contains the similarity matrix, with rows and columns representing different genomic signatures. The values indicate the similarity between the signatures, with higher values representing greater similarity.
  
  ### Example Output Snippet from `aps_comp.csv`
  
  ```
  GCF_002139955_1,GCF_025263725_1,GCF_002019265_2,GCF_017723895_1,GCF_026651895_1,GCF_001610795_1,GCF_000488895_1,GCA_905123925_1,GCF_033441695_1,GCF_001908775_1,22_323
  1.0,0.9265560609248261,0.9977862642066111,0.9889228216483386,0.9073164224987877,0.9700903744261007,0.9889150886270418,0.9115015968590614,0.9041193116358751,0.9045361688901344,0.8871859607482656
  0.9265560609248261,1.0,0.9263988347909979,0.9251119480988551,0.911316733531698,0.9226799924167159,0.9251202135619242,0.9158647097039145,0.9083838785070867,0.9071117431126531,0.8898799125144564
  0.9977862642066111,0.9263988347909979,1.0,0.9888257224992404,0.907831757300304,0.9699598463114608,0.9888096267370786,0.9114840834725453,0.9040737568047027,0.9046135867305022,0.8873565243085494
  0.9889228216483386,0.9251119480988551,0.9888257224992404,1.0,0.9121759558988269,0.970107774713152,0.9999853858427241,0.9135428296802719,0.9038264872056821,0.9051938652156456,0.8894408455767939
  0.9073164224987877,0.911316733531698,0.907831757300304,0.9121759558988269,1.0,0.9037398150211037,0.9120910480325186,0.9296825816700862,0.909749003812823,0.9689599754220248,0.9457379103831718
  0.9700903744261007,0.9226799924167159,0.9699598463114608,0.970107774713152,0.9037398150211037,1.0,0.9701013884045692,0.907211835393209,0.902840929222427,0.9047438469368647,0.8876233079754928
  0.9889150886270418,0.9251202135619242,0.9888096267370786,0.9999853858427241,0.9120910480325186,0.9701013884045692,1.0,0.913551000075958,0.9038346560102811,0.9050856961227841,0.8894454567863859
  0.9115015968590614,0.9158647097039145,0.9114840834725453,0.9135428296802719,0.9296825816700862,0.907211835393209,0.913551000075958,1.0,0.908850448446987,0.9278227730780791,0.9109513479182657
  0.9041193116358751,0.9083838785070867,0.9040737568047027,0.9038264872056821,0.909749003812823,0.902840929222427,0.9038346560102811,0.908850448446987,1.0,0.9082147459073008,0.9050495064323612
  ```

  ### Column Details
  
  - **Signature IDs**: The first row and column contain the IDs of the genomic signatures being compared.
  - **Similarity Values**: The matrix values indicate the similarity between the signatures, with values ranging from 0 to 1. A value of 1 indicates identical signatures, while values closer to 0 indicate less similarity.

  > ## 🧐 What does all this mean? and Why?   
  > 
  >  ### Genomic Signatures and k-mers
  >  
  >  Genomic signatures are unique representations of genomic sequences, often based on k-mers, which capture the unique characteristics of a genome. In this case, a k-mer length of 31 is used.
  >  
  >  ### Average Nucleotide Identity (ANI)
  >  
  >  ANI is a measure of the average nucleotide identity between two genomic sequences, providing a quantitative measure of their similarity. Sourmash computes ANI by comparing the shared k-mers between sequences, calculating the fraction of shared k-mers over the total k-mers in each sequence. Higher ANI values indicate greater similarity between the sequences.
  >  
  >  ### Sourmash
  >  
  >  - **Purpose**: Sourmash is a tool used to compute and compare genomic signatures using k-mers. It is particularly useful for quickly comparing large sets of genomic data.
  >  - **Usage**: Sourmash can handle various types of genomic data, producing similarity matrices that indicate the degree of similarity between different genomes.
  >  - **Output**: The similarity matrix provides a comprehensive view of the relationships between the genomes, which can be used for clustering, phylogenetic analysis, and other comparative studies.
  >  
  >  ### Importance in Pathogen Surveillance
  >  
  >  - **Identifying Similarities**: Comparing genomic signatures helps in identifying similarities and differences between pathogen genomes, which is essential for tracking outbreaks and understanding pathogen evolution.
  >  - **Data Analysis**: The similarity matrix is a key tool for visualizing and interpreting genomic relationships, facilitating better decision-making in pathogen surveillance and research.

</details>

---

This README is designed to be viewed in VSCode Gitpod, with the file explorer on the left, this document on the top, and the terminal on the bottom for an efficient and interactive learning experience.




## Session Leader(s) Bios

<table>
  <tr>
    <td align="center"><strong>Zach Foster</strong><br><img src="content/Untitled.png" width="200"></td>
    <td align="center"><strong>Alex Weisberg</strong><br><img src="content/Untitled%201.png" width="200"></td>
    <td align="center"><strong>Camilo H. Parada Rojas</strong><br><img src="content/Untitled%202.png" width="200"></td>
  </tr>
  <tr>
    <td align="center"><strong>Jeff Chang</strong><br><img src="content/Untitled%203.png" width="200"></td>
    <td align="center"><strong>Martha Sudermann</strong><br><img src="content/Untitled%204.png" width="200"></td>
    <td align="center"><strong>Nik Grunwald</strong><br><img src="content/Untitled%205.png" width="200"></td>
  </tr>
</table>

